src/NUnitEngine/nunit.engine.core/DotNetHelper.cs

```diff
@@ -2,65 +2,120 @@
+                Environment.GetEnvironmentVariable("DOTNET_ROOT_X86") ?? (
+                    OS.IsWindows
+                        ? (string) Registry.LocalMachine.OpenSubKey(X86_SUBKEY1)?.GetValue("InstallLocation") ??
+                          (string) Registry.LocalMachine.OpenSubKey(X86_SUBKEY2)?.GetValue("InstallLocation") ?? @"C:\Program Files (x86)\dotnet"
+                        : "/usr/shared/dotnet/"));
+#pragma warning restore CA1416
+
+        private static Lazy<List<RuntimeInfo>> _x64Runtimes = new Lazy<List<RuntimeInfo>>(() => [.. GetAllRuntimes(x86: false)]);
+        private static Lazy<List<RuntimeInfo>> _x86Runtimes = new Lazy<List<RuntimeInfo>>(() => [.. GetAllRuntimes(x86: true)]);
+
+        public enum Architecture
+        {
+            Unspecified,
+            X64,
+            X86
+        }
 
-        private static string _x64InstallDirectory;
-        public static string GetX64InstallDirectory()
+        public class RuntimeInfo
         {
-            if (_x64InstallDirectory == null)
-                _x64InstallDirectory = Environment.GetEnvironmentVariable("DOTNET_ROOT");
+            public string Name;
+            public Version Version;
+            public string Path;
+
+            public RuntimeInfo(string name, string version, string path)
+                : this(name, new Version(version), path) { }
 
-            if (_x64InstallDirectory == null)
+            public RuntimeInfo(string name, Version version, string path)
             {
-#if NETFRAMEWORK
-                if (Path.DirectorySeparatorChar == '\\')
-#else
-                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-#endif
-                {
-                    RegistryKey key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\dotnet\SetUp\InstalledVersions\x64\sharedHost\");
-                    _x64InstallDirectory = (string)key?.GetValue("Path");
-                }
-                else
-                    _x64InstallDirectory = "/usr/shared/dotnet/";
+                Name = name;
+                Version = version;
+                Path = path;
             }
+        }
+
+        /// <summary>
+        /// Get the correct install directory, depending on whether we need X86 or X64 architecture.
+        /// </summary>
+        /// <param name="x86">Flag indicating whether the X86 architecture is needed</param>
+        /// <returns></returns>
+        public static string GetInstallDirectory(bool x86) => x86
+            ? _x86InstallDirectory.Value : _x64InstallDirectory.Value;
 
-            return _x64InstallDirectory;
+        /// <summary>
+        /// Get the correct dotnet.exe, depending on whether we need X86 or X64 architecture.
+        /// </summary>
+        /// <param name="x86">Flag indicating whether the X86 architecture is needed</param>
+        /// <returns></returns>
+        public static string GetDotnetExecutable(bool x86) => Path.Combine(GetInstallDirectory(x86), "dotnet.exe");
```

Comment:
Not that only windows has a `.exe` suffix, on Linux, the executable is just `dotnet`.

Reply:
I fixed this.